common-ui-compose/src/main/java/app/tivi/common/compose/insets.kt

```diff
@@ -14,64 +14,410 @@
+    // in multi-window, etc.
+    InsetsSizeModifier(insets = DisplayInsetsAmbient.current.systemBars, heightSide = VerticalSide.Top)
+}
+
+/**
+ * Declare the preferred height of the content to match the height of the navigation bar when present at the bottom of the screen.
+ *
+ * This is very handy when used with `Spacer` to push content below the navigation bar:
+ * ```
+ * Column {
+ *     // Content to be drawn above status bar (y-axis)
+ *     Spacer(Modifier.navigationBarHeight())
+ * }
+ * ```
+ *
+ * It's also useful when used to draw a scrim which matches the navigation bar:
+ * ```
+ * Spacer(
+ *     Modifier.navigationBarHeight()
+ *         .fillMaxWidth()
+ *         .drawBackground(MaterialTheme.colors.background.copy(alpha = 0.3f)
+ * )
+ * ```
+ *
+ * Internally this uses [Modifier.height] so has the same characteristics with regards to incoming
+ * layout constraints.
+ */
+fun Modifier.navigationBarHeight() = composed {
+    // TODO: Move to Android 11 WindowInsets APIs when they land in AndroidX.
+    // It currently assumes that nav bar == bottom, which is wrong in landscape.
+    // It also doesn't handle the IME correctly.
+    InsetsSizeModifier(insets = DisplayInsetsAmbient.current.systemBars, heightSide = VerticalSide.Bottom)
+}
+
+enum class HorizontalSide { Left, Right }
+enum class VerticalSide { Top, Bottom }
+
+/**
+ * Declare the preferred width of the content to match the width of the navigation bar,
+ * on the given [side].
+ *
+ * This is very handy when used with `Spacer` to push content inside from any vertical
+ * navigation bars (typically when the device is in landscape):
+ * ```
+ * Row {
+ *     Spacer(Modifier.navigationBarWidth(HorizontalSide.Left))
+ *
+ *     // Content to be inside the navigation bars (x-axis)
+ *
+ *     Spacer(Modifier.navigationBarWidth(HorizontalSide.Right))
+ * }
+ * ```
+ *
+ * It's also useful when used to draw a scrim which matches the navigation bar:
+ * ```
+ * Spacer(
+ *     Modifier.navigationBarWidth(HorizontalSide.Left)
+ *         .fillMaxHeight()
+ *         .drawBackground(MaterialTheme.colors.background.copy(alpha = 0.3f)
+ * )
+ * ```
+ *
+ * Internally this uses [Modifier.width] so has the same characteristics with regards to incoming
+ * layout constraints.
+ *
+ * @param side The navigation bar side to use as the source for the width.
+ */
+fun Modifier.navigationBarWidth(side: HorizontalSide) = composed {
+    // TODO: Move to Android 11 WindowInsets APIs when they land in AndroidX.
+    // It currently assumes that nav bar == left/right
+    InsetsSizeModifier(insets = DisplayInsetsAmbient.current.systemBars, widthSide = side)
+}
+
+/**
+ * Allows conditional setting of [insets] on each dimension.
+ */
+private inline fun Modifier.insetsPadding(
+    insets: Insets,
+    left: Boolean = false,
+    top: Boolean = false,
+    right: Boolean = false,
+    bottom: Boolean = false
+) = this then InsetsPaddingModifier(insets, left, top, right, bottom)
```

Comment:
What about making `InsetsPaddingModifier` just accept Px values for l,t,r,b and converting here? That way the modifier wouldn't need to hold the `insets` object and it would only need to do the `if`s once, not on every measure. Actually you could probably move it to using `Modifier.layout`

Reply:
I'm not sure on this one. This is a micro-optimization, and the whole point of this is to move the check from composition to measure.